### ISSUE TITLE
WaitForEvent in While(..).Do doesn't wait.

### ISSUE THREAD

**Issue by tatejones** [danielgerlag/workflow-core#151]
In this situation the step 'WaitUntilNextLineManagerApproves' returns a WaitForEvent(..).  I would expect the workflow to hold at this step until the subscription receives the event, but it doesn't.  The workflow returns back to the While step.  The WaitForEvent subscription is removed (not sure why) and workflow fails to continue.

In a straight sequence the WaitForEvent will hold, but in a loop it won't.

Has anyone come across this problem?  

.While(request => request.Form.IsSubmitedStatus)
    .Do(then => then
        .StartWith<WaitUntilNextLineManagerApproves>()
            .Input(step => step.Request, request => request)
            .Input(step => step.NumberOfApprovesRequired, request => 2)
            .Output(step => numberOfApprovals, step => step.NextLineManagerLevel)
    )

**Comment by tatejones** [comment]
I am using WaitFor with a DateTime arg and the events are published after the waitfor subscription.
ExecutionResult.WaitForEvent(_context.Workflow.Id, _context.Workflow.Id, DateTime.Now);

**Comment by tatejones** [comment]
Attached is a modified Sample4.
If you uncomment the Console.Write on the MyData. IsStrValueNullOrEmpty() you will see the console fill as it tests the condition.   The thread is so occupied with this check it will *sometimes* miss the subscription / event.   
[WorkflowCore.Sample04.zip](https://github.com/danielgerlag/workflow-core/files/2161767/WorkflowCore.Sample04.zip)

**Comment by danielgerlag** [comment]
That is not accurate, it uses TPL dataflow that has an underlying thread pool.
This workflow seems to execute as expected?  I don't see any issues here.

**Comment by tatejones** [comment]
I was expecting the loop to wait at the WaitFor and only test the condition after the event was received.   When you run the example and enter some values, some events are missed and are never printed to the console.  This won't scale if there hundreds of workflows with a 'while' and 'waitfor'.  It would be become CPU bounded.
    
This is different behaviour for waitfor when not encapsulated within a 'while'.   Outside a 'while' it does wait for an event before progressing to the next step.   Why?

**Comment by tatejones** [comment]
Just running this simple example and doing nothing takes 30% of my CPU.

**Comment by danielgerlag** [comment]
So, the logical workflow in fact does wait, but the while condition continues to be evaluated because of the `DetermineNextExecutionTime` method in the `WorkflowExecutor` class is scheduling it to run again immediately.  We could look at making the `DetermineNextExecutionTime` method smarter for this scenario, so it doesn't schedule it to execute the workflow again.

**Comment by tatejones** [comment]
I agree, as this is a "show stopper" for us.   Somehow the workflow needs to know that there is an encapsulating loop invariant that hasn't changed or the workflow hasn't received or produced any inputs /outputs (eg. external events) that could impact the state of the workflow causing the invariant to change.

Maybe DetermineNextExecution does not move forward unless the workflow overall state hasn't changed.  E.g. external event or the state of the workflow data or step inputs/outputs have changed.  It needs to be simple and efficient.

**Comment by danielgerlag** [comment]
I have deployed a new version to Nuget that should resolve this issue.

**Comment by tatejones** [comment]
thanks.  I will check it out today.
